lib/LLVMIRCodeGen/BundleSaver.cpp

```diff
@@ -450,6 +457,66 @@ void BundleSaver::emitSymbolTable() {
                            irgen_->getBundleName() + "SymbolTable");
 }
 
+void BundleSaver::addBundleObjects(
+    llvm::StringRef bundlePath, const std::vector<std::string> &bundleObjects) {
+
+  // If we do not have extra object files then return early.
+  if (!bundleObjects.size()) {
+    return;
+  }
+
+  // Read original bundle object file as archive member.
+  std::vector<llvm::NewArchiveMember> newMembers;
+  llvm::Expected<llvm::NewArchiveMember> newMember =
+      llvm::NewArchiveMember::getFile(bundlePath.str(),
+                                      /* Deterministic */ true);
+  newMembers.push_back(std::move(*newMember));
+
+  // Add other object files as archive members.
+  for (const auto &objectName : bundleObjects) {
+    // If this object was already added then we skip it.
+    bool objectAdded = false;
+    for (const auto &member : newMembers) {
+      if (member.MemberName.str() == objectName) {
+        objectAdded = true;
+        break;
+      }
+    }
+    if (objectAdded) {
+      continue;
+    }
+    // Find current object and add it as archive member.
+    bool objectFound = false;
+    for (const auto &memBuffRef : bundleObjectRegister) {
```

Comment:
Why does bundleObjectRegister need to be produced before Glow itself is built (if I understand your PR correctly and how it produces BundleObjectRegister.h). Why do you need all this trouble with CMake? Why not read objects file at runtime, when BundleSaver is doing its work? In general, I'm not a big fan of this CMake magic. If there less build-system based solutions, it would be good to hear about them. Ideally, no build system tricks should be needed at all and everything could be done when BundleSaver runs. This way it will be also much easier to customize the behavior through command-line arguments and parameters passed by backends.

Reply:
The intention is to embed in Glow binary some object files which are required for some custom LLVM backends (you can look into the usecase described below) in order to generate a self contained bundle with some extra dependencies which could be hidden to the end-user.
In my opinion there is no much value in having a CLI arg which describes some paths to some **object files external to the Glow tools** because those can be linked as well in the application using ordinary linkers.
The point here is to try to hide some dependencies to the end-user.

Reply:
> I know that on your part is harder to integrate this mechanism because you have a different build system.

It is much harder on our side using Buck and FB internal build configurations. Even without that, I'm still concerned about playing such games with CMake. 

> The point here is to try to hide some dependencies to the end-user.

I'm still not sure why you need so much CMake magic to achieve this "hiding" objective. If you have a pre-defined set of object files you want to always put into the archive for your backend, why not initialize/set the CLI flag with some default paths to your backend-specific/target-specific object files when you initialize your backend, e.g. in the constructor. This way the user does not need to know about this and it is hidden, but does not require any CMake tricks. More over, other backends derived from the LLVMBackend can customize this behavior by providing their own lists of object files, etc. It seems more modular to me. WDYT?

Reply:
@opti-mix I cannot just set the CLI flag to some hard-coded path because I cannot guarantee the file will be at some given location. The CMake magic I do here is 99% the same as what we do with the libjit when serializing the bytecode as arrays and embedding them in the Glow binary. I don't see why we cannot do the same with this also.

Reply:
@mciprian13 

> I cannot just set the CLI flag to some hard-coded path because I cannot guarantee the file will be at some given location. 

Just checking: Would it work if you'd use a relative path (from e.g. some environment variable) instead of an absolute path and then resolve it at runtime? Something like "$LIBJIT_ROOT/objs/myobject1.o". You could then set this environment variable from shell or you could use a command-line flag to set this root directory.

> The CMake magic I do here is 99% the same as what we do with the libjit when serializing the bytecode as arrays and embedding them in the Glow binary. I don't see why we cannot do the same with this also.

I'm not a big fan of this part either ;-) And implementing this libjit serialization into bytecode and then packaging as array was quite a pain to implement using buck :-D

Another question is: IIRC, libjit is compiled by each LLVMBackend-derived backend separately (e.g. by CPUBackend). It is not compiled as part LLVMIRCodeGen. And you PR touches BundleSaver, which is part of the LLVMIRCodeGen. But what if multiple backends are using LLVMIRCodeGen? They should not all use the same BundleSaver with a specific version of the embedded object file. They probably want to use their own versions/sets of object files, which could be target/backend specific.

Reply:
Well if this mechanism is not something you want I'll have to close this PR and merge it internally on our side. I would have preferred for you to except it in one form or another cuz otherwise it will become a pain to maintain it on our side. We need this mechanism ASAP in this exact form.

Just so you know what we do at NXP is not use Glow for ourselves but optimize it and package for others. So adding extra complexity to the usability of the Glow AOT tools (by relying on environment variables, hardcoded paths, etc) is not something we want. With this object archiving mechanism we just want to hide from the user the process of archiving additional objects (compiled with other compilers than clang) to the final object while being totally hidden from the user.

Can we add this mechanism only on the community part of Glow without adding it in your internal build system (if not of interest)? Maybe I can put it under some compile switches so that is build and ran only on github but not internally on your side? This way others might use this feature as well. WDYT?

Reply:
@mciprian13 

> Can we add this mechanism only on the community part of Glow without adding it in your internal build system (if not of interest)? Maybe I can put it under some compile switches so that is build and ran only on github but not internally on your side? This way others might use this feature as well. WDYT?

Yes, I guess this could be an option. 

But even in this case, I'd still highly recommend not changing the LLVMIRCodeGen (i.e. storing the object files under this directory) due to the reasons I described above. Try to move those object files into backends, so that different backends can bundle their own object files. After all, we do a similar trick for libjit, which lives inside backends.